src/mirror/db.rs

```diff
@@ -0,0 +1,808 @@
+            .await?)
+    }
+
+    pub(crate) async fn get_last_active_entry(&self, did: Did) -> anyhow::Result<plc::LogEntry> {
+        let did_saved = did.clone();
+        let entry = self
+            .inner
+            .conn(|conn| match Entry::get_latest_active(conn, did)? {
+                None => Ok(None),
+                Some(entry) => entry.hydrate(conn).map(Some),
+            })
+            .await?;
+
+        if let Some(entry) = entry {
+            Ok(entry.assemble()?)
+        } else {
+            Err(anyhow!("DID not registered: {}", did_saved.as_ref()))
+        }
+    }
+
+    pub(crate) async fn get_audit_log(&self, did: Did) -> anyhow::Result<Vec<plc::LogEntry>> {
+        let entries = self
+            .inner
+            .conn(|conn| {
+                Entry::get_audit_log(conn, did)?
+                    .into_iter()
+                    .map(|entry| entry.hydrate(conn))
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .await?;
+
+        entries.into_iter().map(|entry| entry.assemble()).collect()
+    }
+
+    pub(crate) async fn export(&self, params: ExportParams) -> anyhow::Result<Vec<plc::LogEntry>> {
+        let entries = self
+            .inner
+            .conn(|conn| {
+                Entry::get_log_entries(conn, params.bounded_count(), params.after)?
+                    .into_iter()
+                    .map(|entry| entry.hydrate(conn))
+                    .collect::<Result<Vec<_>, _>>()
+            })
+            .await?;
+
+        entries.into_iter().map(|entry| entry.assemble()).collect()
+    }
+}
+
+const CREATE_DATABASES: &str = "
+CREATE TABLE IF NOT EXISTS identity (
+    identity_id INTEGER PRIMARY KEY,
+    did TEXT NOT NULL UNIQUE
+);
+CREATE TABLE IF NOT EXISTS key (
+    key_id INTEGER PRIMARY KEY,
+    key TEXT NOT NULL UNIQUE
+);
+CREATE TABLE IF NOT EXISTS atproto_pds (
+    pds_id INTEGER PRIMARY KEY,
+    endpoint TEXT NOT NULL UNIQUE
+);
+CREATE TABLE IF NOT EXISTS plc_log (
+    entry_id INTEGER PRIMARY KEY,
+    cid BLOB NOT NULL UNIQUE,
+    identity INTEGER NOT NULL,
+    created_at TEXT,
+    nullified INTEGER,
+    -- operation
+    type TEXT NOT NULL CHECK(type IN ('O','T','C')),
+    also_known_as JSON,
+    atproto_signing INTEGER,
+    atproto_pds INTEGER,
+    prev INTEGER,
+    -- Signatures are stored in their Base64 encoding because
+    -- the log contains signatures with invalid padding.
+    sig TEXT NOT NULL,
+    FOREIGN KEY(identity) REFERENCES identity(identity_id),
+    FOREIGN KEY(atproto_signing) REFERENCES key(key_id)
+    FOREIGN KEY(atproto_pds) REFERENCES atproto_pds(pds_id)
+    FOREIGN KEY(prev) REFERENCES plc_log(entry_id)
+);
+CREATE INDEX IF NOT EXISTS plc_log_idx_created_at ON plc_log(created_at DESC);
+CREATE INDEX IF NOT EXISTS plc_log_idx_identity_created_at ON plc_log(identity, created_at);
+CREATE TABLE IF NOT EXISTS rotation_keys (
+    entry INTEGER NOT NULL,
+    authority INTEGER NOT NULL,
```

Comment:
Changed the database schema to add this explicit field, because otherwise I couldn't guarantee extraction of the correct rotation key order (and was running into canonicity problems for the early legacy create operations).